File: src/semantic_references/lockfile.py
"""``reference-lock-v1``: the generated, checked-in custody observation.

Frozen shape (design spec 0004, section ``reference-lock-v1``):

- schema version, generator identity, entries keyed by source ID;
- each entry: declared origin, tracked ref, resolved ref, full commit and
  tree object IDs plus Git object format, a SHA-256 digest of the complete
  canonical catalog record, retrieval timestamp, acquisition kind, whether
  the origin commit was remotely verified, and per-license-artifact Git
  mode / byte length / raw-blob SHA-256.

The lock is mutated only through :func:`write_lock`, via a temporary file and
atomic replace. Loading is strict: duplicate keys, abbreviated object IDs,
missing fields, unsafe paths, and unknown schema versions are all rejected.
"""

from __future__ import annotations

import contextlib
import json
import os
import re
import tempfile
from dataclasses import dataclass, replace
from pathlib import Path
from typing import Any, cast

from semantic_references.errors import LockFileError

SCHEMA_NAME = "reference-lock-v1"
_OBJECT_FORMAT_HEX_LENGTH = {"sha1": 40, "sha256": 64}
_HEX_RE = {
    "sha1": re.compile(r"^[0-9a-f]{40}$"),
    "sha256": re.compile(r"^[0-9a-f]{64}$"),
}
_SHA256_RE = re.compile(r"^[0-9a-f]{64}$")
_TIMESTAMP_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")
_MODE_RE = re.compile(r"^[0-7]{6}$")
_ACQUISITION_KINDS = frozenset({"remote", "local-sibling", "local-object-cache"})


def _no_duplicate_keys(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise LockFileError(f"duplicate JSON key {key!r}")
        result[key] = value
    return result


_ENTRY_FIELDS = frozenset(
    {
        "origin",
        "track",
        "resolved_ref",
        "object_format",
        "commit",
        "tree",
        "catalog_digest",
        "retrieved_at",
        "acquisition",
        "origin_verified",
        "licenses",
    }
)


def _check_entry_field_set(source_id: str, data: dict[str, object]) -> None:
    missing = _ENTRY_FIELDS - set(data)
    if missing:
        raise LockFileError(f"lock entry {source_id!r} is missing fields: {sorted(missing)}")
    extra = set(data) - _ENTRY_FIELDS
    if extra:
        raise LockFileError(f"lock entry {source_id!r} has unexpected fields: {sorted(extra)}")


def _check_entry_strings(source_id: str, data: dict[str, object]) -> tuple[str, str, str]:
    values: dict[str, str] = {}
    for name in ("origin", "track", "resolved_ref"):
        value = data[name]
        if not isinstance(value, str) or not value:
            raise LockFileError(f"lock entry {source_id!r}: {name!r} must be a non-empty string")
        values[name] = value
    return values["origin"], values["track"], values["resolved_ref"]


def _check_entry_object_ids(source_id: str, data: dict[str, object]) -> tuple[str, str, str]:
    object_format = data["object_format"]
    if not isinstance(object_format, str) or object_format not in _OBJECT_FORMAT_HEX_LENGTH:
        raise LockFileError(
            f"lock entry {source_id!r}: unsupported object_format {object_format!r}"
        )
    hex_re = _HEX_RE[object_format]
    commit = data["commit"]
    if not isinstance(commit, str) or not hex_re.match(commit):
        raise LockFileError(
            f"lock entry {source_id!r}: 'commit' must be a full {object_format} object id"
        )
    tree = data["tree"]
    if not isinstance(tree, str) or not hex_re.match(tree):
        raise LockFileError(
            f"lock entry {source_id!r}: 'tree' must be a full {object_format} object id"
        )
    return object_format, commit, tree


def _check_catalog_digest(source_id: str, catalog_digest: object) -> str:
    if not isinstance(catalog_digest, str) or not _SHA256_RE.match(catalog_digest):
        raise LockFileError(
            f"lock entry {source_id!r}: 'catalog_digest' must be a full sha256 digest"
        )
    return catalog_digest


def _check_retrieved_at(source_id: str, retrieved_at: object) -> str:
    if not isinstance(retrieved_at, str) or not _TIMESTAMP_RE.match(retrieved_at):
        raise LockFileError(
            f"lock entry {source_id!r}: 'retrieved_at' must be an ISO-8601 UTC timestamp"
        )
    return retrieved_at


def _check_acquisition(source_id: str, acquisition: object) -> str:
    if not isinstance(acquisition, str) or acquisition not in _ACQUISITION_KINDS:
        raise LockFileError(
            f"lock entry {source_id!r}: unsupported acquisition kind {acquisition!r}"
        )
    return acquisition


def _check_licenses_field(source_id: str, licenses_raw: object) -> dict[str, LicenseObservation]:
    if not isinstance(licenses_raw, dict) or not licenses_raw:
        raise LockFileError(f"lock entry {source_id!r}: 'licenses' must be a non-empty table")
    typed_raw = cast(dict[str, object], licenses_raw)
    return {path: LicenseObservation.from_json(path, value) for path, value in typed_raw.items()}


@dataclass(frozen=True, slots=True)
class LicenseObservation:
    """A single license artifact's committed-blob measurement."""

    mode: str
    size: int
    sha256: str

    def to_json(self) -> dict[str, object]:
        return {"mode": self.mode, "size": self.size, "sha256": self.sha256}

    @staticmethod
    def from_json(path: str, data: object) -> LicenseObservation:
        if not isinstance(data, dict):
            raise LockFileError(f"license entry {path!r} must be a table")
        typed_data = cast(dict[str, object], data)
        mode = typed_data.get("mode")
        size = typed_data.get("size")
        sha256 = typed_data.get("sha256")
        if not isinstance(mode, str) or not _MODE_RE.match(mode):
            raise LockFileError(f"license entry {path!r}: 'mode' must be a 6-digit octal string")
        if not isinstance(size, int) or isinstance(size, bool) or size < 0:
            raise LockFileError(f"license entry {path!r}: 'size' must be a non-negative integer")
        if not isinstance(sha256, str) or not _SHA256_RE.match(sha256):
            raise LockFileError(f"license entry {path!r}: 'sha256' must be a full 64-hex digest")
        if set(typed_data) != {"mode", "size", "sha256"}:
            raise LockFileError(f"license entry {path!r} has unexpected fields")
        return LicenseObservation(mode=mode, size=size, sha256=sha256)


@dataclass(frozen=True, slots=True)
class LockEntry:
    """One source's recorded custody observation."""

    origin: str
    track: str
    resolved_ref: str
    object_format: str
    commit: str
    tree: str
    catalog_digest: str
    retrieved_at: str
    acquisition: str
    origin_verified: bool
    licenses: dict[str, LicenseObservation]

    def to_json(self) -> dict[str, object]:
        return {
            "origin": self.origin,
            "track": self.track,
            "resolved_ref": self.resolved_ref,
            "object_format": self.object_format,
            "commit": self.commit,
            "tree": self.tree,
            "catalog_digest": self.catalog_digest,
            "retrieved_at": self.retrieved_at,
            "acquisition": self.acquisition,
            "origin_verified": self.origin_verified,
            "licenses": {path: obs.to_json() for path, obs in self.licenses.items()},
        }

    def with_retrieved_at(self, retrieved_at: str) -> LockEntry:
        return replace(self, retrieved_at=retrieved_at)

    def content_equal(self, other: LockEntry) -> bool:
        """Equality ignoring ``retrieved_at`` (used to detect a no-op re-lock)."""
        return self.with_retrieved_at("") == other.with_retrieved_at("")

    @staticmethod
    def from_json(source_id: str, data: object) -> LockEntry:
        if not isinstance(data, dict):
            raise LockFileError(f"lock entry {source_id!r} must be a table")
        typed_data = cast(dict[str, object], data)
        _check_entry_field_set(source_id, typed_data)

        origin, track, resolved_ref = _check_entry_strings(source_id, typed_data)
        object_format, commit, tree = _check_entry_object_ids(source_id, typed_data)
        catalog_digest = _check_catalog_digest(source_id, typed_data["catalog_digest"])
        retrieved_at = _check_retrieved_at(source_id, typed_data["retrieved_at"])
        acquisition = _check_acquisition(source_id, typed_data["acquisition"])
        origin_verified = typed_data["origin_verified"]
        if not isinstance(origin_verified, bool):
            raise LockFileError(f"lock entry {source_id!r}: 'origin_verified' must be a boolean")
        licenses = _check_licenses_field(source_id, typed_data["licenses"])

        return LockEntry(
            origin=origin,
            track=track,
            resolved_ref=resolved_ref,
            object_format=object_format,
            commit=commit,
            tree=tree,
            catalog_digest=catalog_digest,
            retrieved_at=retrieved_at,
            acquisition=acquisition,
            origin_verified=origin_verified,
            licenses=licenses,
        )


@dataclass(frozen=True, slots=True)
class Lock:
    """The full parsed lock file."""

    generator: str
    sources: dict[str, LockEntry]

    def to_json(self) -> dict[str, object]:
        return {
            "schema": SCHEMA_NAME,
            "generator": self.generator,
            "sources": {
                source_id: entry.to_json() for source_id, entry in sorted(self.sources.items())
            },
        }


def serialize_lock(lock: Lock) -> bytes:
    text = json.dumps(lock.to_json(), indent=2, sort_keys=True, ensure_ascii=True)
    return (text + "\n").encode("utf-8")


def parse_lock_text(text: str) -> Lock:
    try:
        parsed: object = json.loads(text, object_pairs_hook=_no_duplicate_keys)
    except json.JSONDecodeError as exc:
        raise LockFileError(f"lock file is not valid JSON: {exc}") from exc

    if not isinstance(parsed, dict):
        raise LockFileError("lock file must be a JSON object")
    document = cast(dict[str, object], parsed)

    expected_top = {"schema", "generator", "sources"}
    missing = expected_top - set(document)
    if missing:
        raise LockFileError(f"lock file is missing top-level fields: {sorted(missing)}")
    extra = set(document) - expected_top
    if extra:
        raise LockFileError(f"lock file has unexpected top-level fields: {sorted(extra)}")

    schema = document["schema"]
    if schema != SCHEMA_NAME:
        raise LockFileError(f"unknown lock schema {schema!r} (expected {SCHEMA_NAME!r})")

    generator = document["generator"]
    if not isinstance(generator, str) or not generator:
        raise LockFileError("lock file 'generator' must be a non-empty string")

    sources_field = document["sources"]
    if not isinstance(sources_field, dict):
        raise LockFileError("lock file 'sources' must be a JSON object")
    sources_raw = cast(dict[str, object], sources_field)

    sources: dict[str, LockEntry] = {}
    for source_id, entry_data in sources_raw.items():
        sources[source_id] = LockEntry.from_json(source_id, entry_data)

    return Lock(generator=generator, sources=sources)


def load_lock(path: Path) -> Lock:
    try:
        text = path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return Lock(generator="", sources={})
    except OSError as exc:
        raise LockFileError(f"cannot read lock file at {path}: {exc}") from exc
    return parse_lock_text(text)


def write_lock(path: Path, lock: Lock) -> None:
    """Atomically replace the lock file with ``lock``'s canonical bytes."""
    new_bytes = serialize_lock(lock)
    try:
        existing_bytes = path.read_bytes()
    except FileNotFoundError:
        existing_bytes = None
    if existing_bytes == new_bytes:
        return

    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(prefix=".sources.lock.", suffix=".tmp", dir=path.parent)
    try:
        with os.fdopen(fd, "wb") as handle:
            handle.write(new_bytes)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp_name, path)
    except BaseException:
        with contextlib.suppress(OSError):
            os.unlink(tmp_name)
        raise

File: src/references/lockfile.ts
import { Effect, FileSystem, Path } from "effect";
import { stringifyCanonicalJson } from "./canonical-json.ts";
import {
  isConcreteGitRef,
  isGitSafeValue,
  validateLicensePath,
  validateSourceId,
} from "./catalog.ts";
import { LockFileError } from "./errors.ts";
import { parseStrictJson } from "./strict-json.ts";

export const SCHEMA_NAME = "reference-lock-v1";

const OBJECT_FORMAT_HEX_LENGTH: Readonly<Record<string, number>> = { sha1: 40, sha256: 64 };
const HEX_PATTERN: Readonly<Record<string, RegExp>> = {
  sha1: /^[0-9a-f]{40}$/,
  sha256: /^[0-9a-f]{64}$/,
};
const SHA256_PATTERN = /^[0-9a-f]{64}$/;
const TIMESTAMP_PATTERN = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
const REMOTE_ACQUISITION = "remote";
const ACQUISITION_KINDS = new Set([REMOTE_ACQUISITION, "local-sibling", "local-object-cache"]);
/** Git blob modes for a regular file: excludes symlinks (120000) and gitlinks (160000). */
const REGULAR_BLOB_MODES = new Set(["100644", "100755"]);

const ENTRY_FIELDS = new Set([
  "origin",
  "track",
  "resolved_ref",
  "object_format",
  "commit",
  "tree",
  "catalog_digest",
  "retrieved_at",
  "acquisition",
  "origin_verified",
  "licenses",
]);
const LICENSE_FIELDS = new Set(["mode", "size", "sha256"]);
const TOP_LEVEL_FIELDS = new Set(["schema", "generator", "sources"]);

export interface LicenseObservation {
  readonly mode: string;
  readonly size: number | bigint;
  readonly sha256: string;
}

export interface LockEntry {
  readonly origin: string;
  readonly track: string;
  readonly resolvedRef: string;
  readonly objectFormat: string;
  readonly commit: string;
  readonly tree: string;
  readonly catalogDigest: string;
  readonly retrievedAt: string;
  readonly acquisition: string;
  readonly originVerified: boolean;
  readonly licenses: ReadonlyMap<string, LicenseObservation>;
}

export interface Lock {
  readonly generator: string;
  readonly sources: ReadonlyMap<string, LockEntry>;
}

const licenseObservationToJson = (observation: LicenseObservation): Record<string, unknown> => ({
  mode: observation.mode,
  size: observation.size,
  sha256: observation.sha256,
});

const lockEntryToJson = (entry: LockEntry): Record<string, unknown> => ({
  origin: entry.origin,
  track: entry.track,
  resolved_ref: entry.resolvedRef,
  object_format: entry.objectFormat,
  commit: entry.commit,
  tree: entry.tree,
  catalog_digest: entry.catalogDigest,
  retrieved_at: entry.retrievedAt,
  acquisition: entry.acquisition,
  origin_verified: entry.originVerified,
  licenses: Object.fromEntries(
    [...entry.licenses].map(([path, observation]) => [path, licenseObservationToJson(observation)]),
  ),
});

export const lockToJson = (lock: Lock): Record<string, unknown> => ({
  schema: SCHEMA_NAME,
  generator: lock.generator,
  sources: Object.fromEntries(
    [...lock.sources].map(([sourceId, entry]) => [sourceId, lockEntryToJson(entry)]),
  ),
});

export const serializeLock = (lock: Lock): Uint8Array =>
  new TextEncoder().encode(`${stringifyCanonicalJson(lockToJson(lock), 2)}\n`);

export const lockEntryContentEqual = (left: LockEntry, right: LockEntry): boolean =>
  stringifyCanonicalJson(lockEntryToJson({ ...left, retrievedAt: "" })) ===
  stringifyCanonicalJson(lockEntryToJson({ ...right, retrievedAt: "" }));

const bytesEqual = (left: Uint8Array, right: Uint8Array): boolean =>
  left.length === right.length && left.every((byte, index) => byte === right[index]);

const isPlainObject = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const requireExactFields = (
  label: string,
  data: Readonly<Record<string, unknown>>,
  expected: ReadonlySet<string>,
): Effect.Effect<void, LockFileError> => {
  const present = new Set(Object.keys(data));
  const missing = [...expected].filter((key) => !present.has(key)).sort();
  if (missing.length > 0) {
    return Effect.fail(
      new LockFileError({ message: `${label} is missing fields: ${JSON.stringify(missing)}` }),
    );
  }
  const extra = [...present].filter((key) => !expected.has(key)).sort();
  if (extra.length > 0) {
    return Effect.fail(
      new LockFileError({ message: `${label} has unexpected fields: ${JSON.stringify(extra)}` }),
    );
  }
  return Effect.void;
};

const requireGitSafeString = (
  label: string,
  value: unknown,
): Effect.Effect<string, LockFileError> =>
  typeof value === "string" && value.length > 0 && isGitSafeValue(value)
    ? Effect.succeed(value)
    : Effect.fail(
        new LockFileError({
          message: `${label} must be a non-empty string that is safe (option-like or has control characters)`,
        }),
      );

const requireBoolean = (label: string, value: unknown): Effect.Effect<boolean, LockFileError> =>
  typeof value === "boolean"
    ? Effect.succeed(value)
    : Effect.fail(new LockFileError({ message: `${label} must be a boolean` }));

const requireNonNegativeInteger = (
  label: string,
  value: unknown,
): Effect.Effect<bigint, LockFileError> =>
  typeof value === "bigint" && value >= 0n
    ? Effect.succeed(value)
    : Effect.fail(new LockFileError({ message: `${label} must be a non-negative integer` }));

const validateLicenseObservation = (
  path: string,
  data: unknown,
): Effect.Effect<LicenseObservation, LockFileError> =>
  Effect.gen(function* () {
    if (!isPlainObject(data)) {
      return yield* new LockFileError({
        message: `license entry ${JSON.stringify(path)} must be a table`,
      });
    }
    yield* requireExactFields(`license entry ${JSON.stringify(path)}`, data, LICENSE_FIELDS);

    const mode = data.mode;
    if (typeof mode !== "string" || !REGULAR_BLOB_MODES.has(mode)) {
      return yield* new LockFileError({
        message:
          `license entry ${JSON.stringify(path)}: 'mode' must be a regular blob mode ` +
          `${JSON.stringify([...REGULAR_BLOB_MODES].sort())} — a symlink, gitlink, or tree cannot ` +
          "be a license artifact",
      });
    }
    const size = yield* requireNonNegativeInteger(
      `license entry ${JSON.stringify(path)}: 'size'`,
      data.size,
    );
    const sha256 = data.sha256;
    if (typeof sha256 !== "string" || !SHA256_PATTERN.test(sha256)) {
      return yield* new LockFileError({
        message: `license entry ${JSON.stringify(path)}: 'sha256' must be a full 64-hex digest`,
      });
    }
    return { mode, size, sha256 };
  });

const validateLicensesField = (
  sourceId: string,
  data: unknown,
): Effect.Effect<ReadonlyMap<string, LicenseObservation>, LockFileError> =>
  Effect.gen(function* () {
    if (!isPlainObject(data) || Object.keys(data).length === 0) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'licenses' must be a non-empty table`,
      });
    }
    const licenses = new Map<string, LicenseObservation>();
    for (const [path, value] of Object.entries(data)) {
      const validPath = yield* validateLicensePath(path).pipe(
        Effect.mapError(
          (cause) =>
            new LockFileError({
              message: `lock entry ${JSON.stringify(sourceId)}: unsafe license path: ${cause.message}`,
              cause,
            }),
        ),
      );
      const observation = yield* validateLicenseObservation(validPath, value);
      licenses.set(validPath, observation);
    }
    return licenses;
  });

const validateLockEntry = (
  sourceId: string,
  data: unknown,
): Effect.Effect<LockEntry, LockFileError> =>
  Effect.gen(function* () {
    if (!isPlainObject(data)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)} must be a table`,
      });
    }
    yield* requireExactFields(`lock entry ${JSON.stringify(sourceId)}`, data, ENTRY_FIELDS);

    const origin = yield* requireGitSafeString(
      `lock entry ${JSON.stringify(sourceId)}: 'origin'`,
      data.origin,
    );
    const track = yield* requireGitSafeString(
      `lock entry ${JSON.stringify(sourceId)}: 'track'`,
      data.track,
    );
    const resolvedRef = yield* requireGitSafeString(
      `lock entry ${JSON.stringify(sourceId)}: 'resolved_ref'`,
      data.resolved_ref,
    );
    if (!isConcreteGitRef(resolvedRef)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'resolved_ref' must be a concrete valid refs/... name`,
      });
    }

    const objectFormat = data.object_format;
    if (typeof objectFormat !== "string" || !(objectFormat in OBJECT_FORMAT_HEX_LENGTH)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: unsupported object_format ${JSON.stringify(objectFormat)}`,
      });
    }
    const hexPattern = HEX_PATTERN[objectFormat]!;
    const commit = data.commit;
    if (typeof commit !== "string" || !hexPattern.test(commit)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'commit' must be a full ${objectFormat} object id`,
      });
    }
    const tree = data.tree;
    if (typeof tree !== "string" || !hexPattern.test(tree)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'tree' must be a full ${objectFormat} object id`,
      });
    }

    const catalogDigest = data.catalog_digest;
    if (typeof catalogDigest !== "string" || !SHA256_PATTERN.test(catalogDigest)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'catalog_digest' must be a full sha256 digest`,
      });
    }

    const retrievedAt = data.retrieved_at;
    if (typeof retrievedAt !== "string" || !TIMESTAMP_PATTERN.test(retrievedAt)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: 'retrieved_at' must be an ISO-8601 UTC timestamp`,
      });
    }

    const acquisition = data.acquisition;
    if (typeof acquisition !== "string" || !ACQUISITION_KINDS.has(acquisition)) {
      return yield* new LockFileError({
        message: `lock entry ${JSON.stringify(sourceId)}: unsupported acquisition kind ${JSON.stringify(acquisition)}`,
      });
    }
    const originVerified = yield* requireBoolean(
      `lock entry ${JSON.stringify(sourceId)}: 'origin_verified'`,
      data.origin_verified,
    );
    // Only a remote acquisition can have observed the declared origin; a
    // local sibling or object cache is an assumption about identity, never a
    // verification of it, and the converse pairing is equally impossible.
    if (originVerified !== (acquisition === REMOTE_ACQUISITION)) {
      return yield* new LockFileError({
        message:
          `lock entry ${JSON.stringify(sourceId)}: acquisition ${JSON.stringify(acquisition)} cannot report ` +
          `'origin_verified' ${originVerified}`,
      });
    }

    const licenses = yield* validateLicensesField(sourceId, data.licenses);

    const entry: LockEntry = {
      origin,
      track,
      resolvedRef,
      objectFormat,
      commit,
      tree,
      catalogDigest,
      retrievedAt,
      acquisition,
      originVerified,
      licenses,
    };
    return entry;
  });

export const parseLockText = (text: string): Effect.Effect<Lock, LockFileError> =>
  Effect.gen(function* () {
    const parsed = yield* parseStrictJson(text).pipe(
      Effect.mapError(
        (cause) =>
          new LockFileError({ message: `lock file is not valid JSON: ${cause.message}`, cause }),
      ),
    );
    if (!isPlainObject(parsed)) {
      return yield* new LockFileError({ message: "lock file must be a JSON object" });
    }
    yield* requireExactFields("lock file", parsed, TOP_LEVEL_FIELDS);

    if (parsed.schema !== SCHEMA_NAME) {
      return yield* new LockFileError({
        message: `unknown lock schema ${JSON.stringify(parsed.schema)} (expected ${JSON.stringify(SCHEMA_NAME)})`,
      });
    }
    const generator = parsed.generator;
    if (typeof generator !== "string" || generator.length === 0) {
      return yield* new LockFileError({
        message: "lock file 'generator' must be a non-empty string",
      });
    }
    if (!isPlainObject(parsed.sources)) {
      return yield* new LockFileError({
        message: "lock file 'sources' must be a JSON object",
      });
    }

    const sources = new Map<string, LockEntry>();
    for (const [sourceId, entryData] of Object.entries(parsed.sources)) {
      yield* validateSourceId(sourceId).pipe(
        Effect.mapError(
          (cause) =>
            new LockFileError({
              message: `lock file source id ${JSON.stringify(sourceId)} is unsafe: ${cause.message}`,
              cause,
            }),
        ),
      );
      const entry = yield* validateLockEntry(sourceId, entryData);
      sources.set(sourceId, entry);
    }
    return { generator, sources };
  });

const EMPTY_LOCK: Lock = { generator: "", sources: new Map() };

export const loadLock = (path: string): Effect.Effect<Lock, LockFileError, FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const exists = yield* fs
      .exists(path)
      .pipe(
        Effect.mapError(
          (cause) => new LockFileError({ message: `cannot inspect lock file at ${path}`, cause }),
        ),
      );
    if (!exists) return EMPTY_LOCK;
    const text = yield* fs
      .readFileString(path)
      .pipe(
        Effect.mapError(
          (cause) => new LockFileError({ message: `cannot read lock file at ${path}`, cause }),
        ),
      );
    return yield* parseLockText(text);
  });

/**
 * Atomically replace a lock with canonical bytes. A byte-identical lock is a
 * true no-op; otherwise the new file is written and synced in the destination
 * directory before one rename makes it visible. The scoped temporary file is
 * removed on every pre-rename failure.
 */
export const writeLock = (
  path: string,
  lock: Lock,
): Effect.Effect<void, LockFileError, FileSystem.FileSystem | Path.Path> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const paths = yield* Path.Path;
    const bytes = serializeLock(lock);
    const mapWriteError = (cause: unknown) =>
      new LockFileError({ message: `cannot atomically write lock file at ${path}`, cause });

    const exists = yield* fs.exists(path).pipe(Effect.mapError(mapWriteError));
    if (exists) {
      const current = yield* fs.readFile(path).pipe(Effect.mapError(mapWriteError));
      if (bytesEqual(current, bytes)) return;
    }

    const directory = paths.dirname(path);
    yield* fs.makeDirectory(directory, { recursive: true }).pipe(Effect.mapError(mapWriteError));
    yield* Effect.scoped(
      Effect.gen(function* () {
        const temporary = yield* fs.makeTempFileScoped({
          directory,
          prefix: ".sources.lock.",
          suffix: ".tmp",
        });
        yield* Effect.scoped(
          Effect.gen(function* () {
            const file = yield* fs.open(temporary, { flag: "w" });
            yield* file.writeAll(bytes);
            yield* file.sync;
          }),
        );
        yield* fs.rename(temporary, path);
      }),
    ).pipe(Effect.mapError(mapWriteError));
  });
